drgn/helpers/linux/idr.py

```diff
@@ -30,9 +110,54 @@ def idr_for_each(idr: Object) -> Iterator[Tuple[int, Object]]:
     :param idr: ``struct idr *``
     :return: Iterator of (index, ``void *``) tuples.
     """
-    try:
-        base = idr.idr_base.value_()
-    except AttributeError:
-        base = 0
-    for index, entry in radix_tree_for_each(idr.idr_rt.address_of_()):
-        yield index + base, entry
+    if idr.prog_.type("struct idr").has_member("idr_rt"):
+        try:
+            base = idr.idr_base.value_()
+        except AttributeError:
+            base = 0
+        for index, entry in radix_tree_for_each(idr.idr_rt.address_of_()):
+            yield index + base, entry
+    else:
+        # kernels < 4.11 don't use radix_tree as idr backend
+        prog = idr.prog_
+        next_id = 0
+
+        def _idr_get_next(tmp_id: int) -> Tuple[int, Object]:
+            pa = [ids for ids in range(_max_idr_level(prog) + 1)]
+
+            pa[0] = idr.top.value_()
+            if pa[0] == 0:
+                return 0, NULL(prog, "void *")
+
+            pa_idx = 0
+            idr_layer = Object(prog, "struct idr_layer", address=pa[0])
+            n = (idr_layer.layer.value_() + 1) * IDR_BITS
+            max_id = _idr_max(prog, idr_layer.layer.value_() + 1)
+            while tmp_id >= 0 and tmp_id <= max_id:
+                p = pa[pa_idx]
+                while n > 0 and p:
+                    n = n - IDR_BITS
+                    ary_index = (tmp_id >> n) & IDR_MASK
+                    p = Object(prog, "struct idr_layer", address=p).ary.value_()[
+                        ary_index
+                    ]
+                    pa_idx = pa_idx + 1
+                    pa[pa_idx] = p
+
+                if p != 0:
+                    return tmp_id, Object(prog, "void *", p)
+
+                tmp_id = _round_up(tmp_id + 1, (1 << n))
+
+                while n < _fls(tmp_id):
+                    n = n + IDR_BITS
+                    pa_idx = pa_idx - 1
+
+            return tmp_id, NULL(prog, "void *")
+
+        next_id = 0
+        next_id, entry = _idr_get_next(next_id)
+        while entry != NULL(prog, "void *"):
+            yield next_id, entry
+            next_id = next_id + 1
+            next_id, entry = _idr_get_next(next_id)
```

Comment:
Doing this recursively is simpler and likely more efficient since it will require fewer reads from the core dump. Something like:

```suggestion
        voidp_type = idr.prog_.type("void *")

        def aux(p: Object, id: int, n: int) -> Iterator[Tuple[int, Object]]:
            p = p.read_()
            if p:
                if n == 0:
                    yield id, cast(voidp_type, p)
                else:
                    n -= IDR_BITS
                    for child in p.ary:
                        yield from aux(child, id, n)
                        id += 1 << n

        yield from aux(idr.top, 0, idr.layers.value_() * IDR_BITS)
```

Reply:
Indeed. Thanks for doing this.